### ISSUE TITLE
support namespace option in vault resources

### ISSUE THREAD

**Issue by venky999** [hashicorp/terraform-provider-vault#546]
At the moment vault provider for terraform supports passing namespace in provider config...

provider "vault" {
...
namespace = "test"
}

but If I have a scenario where I have two namespace

- root
- test

and in root namespace I create a PKI root(assume it is self signed for now)
and in test namespace I create PKI-INT and create a CSR and send the request to root namespace and get the signed cert and upload to PKI-INT in test namespace..

In this case its hard to do the above using terraform vault pki resources...suppose If I  make a module for above steps using vault pki  resources..

I can only pass one provider to the module ( that means all the calls are happen in namespace specified in the provider)

I will get error something like this

```
URL: POST https://vault-xxxxxxxxxxx/v1/sys/mounts/test/pki-int
Code: 400. Errors:

* given mount path is not in the same namespace as the request
```

It would be great if we can pass namespace to the resources it self... like the vault rest API

**Comment by malnick** [comment]
Thanks for raising this @venky999. Currently we only support using one namespace per Terraform run, https://github.com/terraform-providers/terraform-provider-vault/blob/master/vault/provider.go#L117. I'm going to recommend this feature for our upcoming release cycle.

**Comment by vmignot** [comment]
Hi, 

I stumble upon the same issue: have you tried to pass two provider to the module as described [here](https://www.terraform.io/docs/configuration/modules.html#passing-providers-explicitly)?

```
provider "vault" {
  alias   = "root"
  version = "~> 2.3"
}

provider "vault" {
  namespace = "test"
  alias     = "test"
  version   = "~> 2.3"
}

module "test" {
  source = "../modules/pki"

  namespace_name = "test"

  providers = {
    vault.root = "vault.root"
    vault      = "vault.test" # Will be default
  }
}
```

Then, in the module itself, you need to declare the one with an alias:
```
provider "vault" {
  alias = "root"
}
```
Then you can specify whenever you need `provider = "vault.root"` and use `depends_on` to enforce the good execution order.

Hope that helps!

**Comment by venky999** [comment]
thank you @vmignot .. that works

**Comment by seeafish** [comment]
@vmignot That is also how I managed to use it. However, for our setup we wanted several levels of namespaces created and would then need 4-5 or even more providers potentially. It's a good workaround for small cases, but for more complex namespace structures, we really need to have the namespace limiter on the resource as per the Vault api.

**Comment by mrpaws** [comment]
Did this ever get merged in?  This would be really nice for our use case in using Terraform to manage vault with several namespaces.

**Comment by m0ps** [comment]
Solution advised by @vmignot works fine in case subnamespace already created, but... if you need to create subnamespace and then create a resource in it - fails. The provider tried to obtain sub-token prior to creating a sub-namespace and fail with 404 (as expected, cuz there is no subnamespace already exist).
So... namespace option for each resource is still needed in most cases.

**Comment by catsby** [comment]
Hey @m0ps  - I apologize for the long delay in response. I'm curious if you could provide an example configuration that demonstrates the scenario you describe? Reading the scenario you describe, I _believe_ specifying a `depends_on` relationship could achieve what you're looking for. 

Example
```
// the default provider definition, the "root" namespace
provider "vault" {}

resource "vault_namespace" "ns1" {
  path = "ns1"
}

provider "vault" {
  alias = "ns1"
  namespace ="ns1"
}

// the nested namespace. We use depends_on to ensure the parent namespace, ns1, exists first. We must use the ns1 provider alias because the "path" attribute of vault_namespace cannot have a "/" in it
resource "vault_namespace" "ns2" {
  provider = vault.ns1
  depends_on=[vault_namespace.ns1]
  path = "ns2"
}
```

Let us know if that's not quite what you were thinking.

**Comment by m0ps** [comment]
Hi @catsby 
depends_on works fine in case of subnamespace already created. But... If you are going to create ns2 in ns1 (which is already creates) and then create a policy in ns2, you need to configure the vault provider two times (for each namespace). In such a case, terraform will try to obtain credentials for both namespaces before ns2 is created and fails.

**Comment by non7top** [comment]
Would be great to have support for this feature as it will work better with `for_each` and modules. Atm you can't iterate through a list of namespaces with vault as you can do with k8s provider where you specify namespace in each resource.

**Comment by devhopsde** [comment]
+1

**Comment by waxb** [comment]
+1

**Comment by wagnst** [comment]
Is there any update on this one? This would be definately needed.

**Comment by waxb** [comment]
+1

**Comment by GoldbergAlexander** [comment]
+100

**Comment by benashz** [comment]
Closed with #1305
